File: app/src/main/java/com/alvarlagerlof/koda/Projects/ProjectsFragment.java
package com.alvarlagerlof.koda.Projects;

import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.alvarlagerlof.koda.DividerItemDecoration;
import com.alvarlagerlof.koda.R;
import com.alvarlagerlof.koda.Utils.DateConversionUtils;

import java.util.ArrayList;

import io.realm.Realm;
import io.realm.RealmChangeListener;
import io.realm.RealmResults;
import io.realm.Sort;

/**
 * Created by alvar on 2016-07-02.
 */

public class ProjectsFragment extends Fragment {

    ProjectsAdapter adapter;
    ArrayList<Object> projectsList = new ArrayList<>();


    @Override
    public View onCreateView(final LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        // Init
        View view = inflater.inflate(R.layout.projects_fragment, container, false);


        // Recyclerview
        adapter = new ProjectsAdapter(projectsList, getActivity().getSupportFragmentManager());
        RecyclerView recyclerView = (RecyclerView) view.findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.addItemDecoration(new DividerItemDecoration(getContext()));
        recyclerView.setAdapter(adapter);


        // On FAB click
        FloatingActionButton floatingActionButton = (FloatingActionButton) view.findViewById(R.id.floatingActionButton);
        floatingActionButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                new ProjectsAddDialog(getContext());
            }
        });


        // Get data once
        getData();

        // On database change
        Realm.getDefaultInstance().addChangeListener(new RealmChangeListener<Realm>() {
            @Override
            public void onChange(Realm element) {
                getData();
            }
        });





        return view;
    }




    public void getData() {

        // Clear adapter list
        projectsList.clear();


        // Get all objects and add to list
        RealmResults<ProjectsRealmObject> realmObjects = Realm.getDefaultInstance().where(ProjectsRealmObject.class).findAll()
                .sort("updatedRealm", Sort.DESCENDING);
        for (int i = 0; i < realmObjects.size(); i++) {

            String privateID    = realmObjects.get(i).getprivateID();
            String publicID     = realmObjects.get(i).getpublicID();
            String title        = realmObjects.get(i).getTitle();
            String updated      = DateConversionUtils.convert(realmObjects.get(i).getUpdatedRealm());
            String description  = realmObjects.get(i).getDescription();
            boolean isPublic    = realmObjects.get(i).getIsPublic();
            String code         = realmObjects.get(i).getCode();
            projectsList.add(new ProjectsObjectData(privateID, publicID, title, updated, description, isPublic, code));
        }



        // If no objects
        if (realmObjects.size() == 0) {
            projectsList.add(new ProjectsObjectNoItems());
        }


        // Notify adapter
        adapter.notifyDataSetChanged();
    }


}

File: app/src/main/java/com/alvarlagerlof/koda/Projects/ProjectsFragment.kt
package com.alvarlagerlof.koda.Projects

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.alvarlagerlof.koda.DividerItemDecoration
import com.alvarlagerlof.koda.Extensions.animateFadeIn
import com.alvarlagerlof.koda.Extensions.animateFadeOut
import com.alvarlagerlof.koda.Extensions.inflate
import com.alvarlagerlof.koda.Extensions.timeStampToDate
import com.alvarlagerlof.koda.Main.MainElevationEvent
import com.alvarlagerlof.koda.R
import com.arasthel.asyncjob.AsyncJob
import io.realm.Realm
import io.realm.RealmChangeListener
import io.realm.Sort
import kotlinx.android.synthetic.main.projects_fragment.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import java.util.*


/**
 * Created by alvar on 2016-07-02.
 */

class ProjectsFragment : Fragment() {

    private var syncVisible = false
    private var syncCounter = 1000000000


    internal lateinit var adapter: ProjectsAdapter
    internal var projectsList: ArrayList<Any> = ArrayList()






    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return container?.inflate(R.layout.projects_fragment)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        EventBus.getDefault().post(MainElevationEvent(elevation = 16f))


        showSync()

        adapter = ProjectsAdapter(projectsList, activity.supportFragmentManager)
        recycler_view.layoutManager = LinearLayoutManager(context)
        recycler_view.addItemDecoration(DividerItemDecoration(context))
        recycler_view.adapter = adapter



        // Realm on change
        val realm = Realm.getDefaultInstance()
        val realmListener = RealmChangeListener<Realm> {
            showSync()
            syncCounter = 1000
        }
        realm.addChangeListener(realmListener)


        // Sync counter
        AsyncJob.doInBackground {
            while (true) {
                try { Thread.sleep(100) } catch (ie: InterruptedException) {}
                if (syncCounter >= 0) {
                    syncCounter -= 100
                    if (syncCounter <= 0) {
                        AsyncJob.doOnMainThread { loadList() }
                    }
                }
            }
        }


        fab?.setOnClickListener {
            ProjectsAddDialog(context)
        }
    }





    fun loadList() {

        showSync()

        // Clear adapter list
        projectsList.clear()


        // Get all objects and add to list
        val realmObjects = Realm.getDefaultInstance()
                .where(ProjectsRealmObject::class.java)
                .findAll()
                .sort("updatedRealm", Sort.DESCENDING)

        for (realmObject in realmObjects) {
            projectsList.add(ProjectsObjectData(
                    realmObject.privateID,
                    realmObject.publicID,
                    realmObject.title,
                    realmObject.updatedRealm.timeStampToDate(),
                    realmObject.description,
                    realmObject.isPublic,
                    realmObject.code))
        }


        // If no objects
        if (realmObjects.size == 0) {
            projectsList.add(ProjectsObjectNoItems())
        }


        // Notify adapter
        adapter.notifyDataSetChanged()

        if (recycler_view != null) {
            recycler_view.smoothScrollToPosition(0)
        }

        hideSync()
    }



    fun showSync() {
        if (!syncVisible) {
            if (sync_view != null) {
                sync_view.visibility = View.VISIBLE
                sync_view.animateFadeIn()
                syncVisible = true
            }
        }
    }

    fun hideSync() {
        if (syncVisible) {
            if (sync_view != null) {
                sync_view.visibility = View.GONE
                sync_view.animateFadeOut()
                syncVisible = false
            }
        }

    }



    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMessageEvent(@Suppress("UNUSED_PARAMETER") event: ProjectsSyncEvent) {
        when (event.message) {
            "show" -> showSync()
            "hide" -> hideSync()
            "offline" -> loadList()
        }
        loadList()
    }




    override fun onStart() {
        super.onStart()
        EventBus.getDefault().register(this)
    }

    override fun onStop() {
        super.onStop()
        EventBus.getDefault().unregister(this)
    }


}
